### PR TITLE
AdSense/Doubleclick Fast Fetch integration with amp-consent

### PR DESCRIPTION
**Pull request by keithwrightbos**
Integrates AdSense/Doubleclick Fast Fetch with amp-consent flow as follows:

- for INSUFFICIENT state, npa=1 is included on ad request indicating non-personalized ad flow should be used
- will not send ad request for UNKNOWN state unless data-npa-on-unknown-consent=true is set on amp-ad element
- RTC is not sent for INSUFFICIENT or UNKNOWN state

Documentation to be added in subsequent PR

### CODE REVIEW

File: extensions/amp-a4a/0.1/real-time-config-manager.js [hunk 1]
```diff
@@ -127,14 +128,24 @@ export function getCalloutParam_(url) {
...
-export function maybeExecuteRealTimeConfig_(a4aElement, customMacros) {
+export function maybeExecuteRealTimeConfig_(
+  a4aElement, customMacros, consentState) {
   const rtcConfig = validateRtcConfig_(a4aElement.element);
   if (!rtcConfig) {
     return;
   }
+  // TODO(keithwrightbos) - allow pub to override such that some callouts are
+  // still allowed.
+  if (consentState == CONSENT_POLICY_STATE.INSUFFICIENT ||
```
[the remark this answers is not in the document]
Comment (keithwrightbos) [reply extensions/amp-a4a/0.1/real-time-config-manager.js]: Done

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 2]
```diff
@@ -658,12 +662,21 @@ export class AmpA4A extends AMP.BaseElement {
...
           }
         })
+        // Possibly block on amp-consent.
+        /** @return {!Promise<?CONSENT_POLICY_STATE>} */
+        .then(() => {
+          checkStillCurrent();
+          const consentPolicyId = super.getConsentPolicy();
+          return consentPolicyId ?
+            getConsentPolicyState(this.getAmpDoc(), consentPolicyId) :
+            Promise.resolve();
```
Comment (zhouyx) [inline extensions/amp-a4a/0.1/amp-a4a.js]: nit: `Promise.resolve(null)`

File: extensions/amp-a4a/0.1/amp-a4a.js [hunk 3]
```diff
@@ -1136,10 +1149,11 @@ export class AmpA4A extends AMP.BaseElement {
   /**
    * Gets the Ad URL to send an XHR Request to.  To be implemented
    * by network.
+   * @param {?CONSENT_POLICY_STATE} unusedConsentState
    * @param {Promise<!Array<rtcResponseDef>>=} opt_rtcResponsesPromise
    * @return {!Promise<string>|string}
    */
-  getAdUrl(opt_rtcResponsesPromise) {
+  getAdUrl(unusedConsentState, opt_rtcResponsesPromise) {
```
Comment (zhouyx) [inline extensions/amp-a4a/0.1/amp-a4a.js]: just to confirm only doubleclick use `opt_rtcResponsesPromise`?

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js [hunk 4]
```diff
@@ -592,16 +604,28 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
...
+  }
+
+  /** @override */
+  getAdUrl(consentState, opt_rtcResponsesPromise) {
+    if (consentState == CONSENT_POLICY_STATE.UNKNOWN &&
+        this.element.getAttribute('data-npa-on-unknown-consent') != 'true') {
+      user().info(TAG, 'Ad request suppressed due to unknown consent');
+      return Promise.resolve('');
+    }
     if (this.iframe && !this.isRefreshing) {
```
Comment (zhouyx) [inline extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js]: Not familiar with the refreshing code, but I think we need to check for  `getConsentPolicyState` before every ad request.

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js [hunk 5]
```diff
@@ -273,6 +286,8 @@ export class AmpAdNetworkAdsenseImpl extends AmpA4A {
       'w': this.size_.width,
       'h': this.size_.height,
       'iu': slotname,
+      'npa': consentState == CONSENT_POLICY_STATE.INSUFFICIENT ||
+          consentState == CONSENT_POLICY_STATE.UNKNOWN ? 1 : null,
```
Comment (zhouyx) [inline extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: I remember we want to distinguish between `SUFFICIENT` and  `UNKNOWN_NOT_REQUIRED`. Do we have any plan on passing the value?

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 2 above)
Comment (lannka) [inline extensions/amp-a4a/0.1/amp-a4a.js]: Promise.resolve(null);

Comment (keithwrightbos) [reply extensions/amp-a4a/0.1/amp-a4a.js]: Done

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 3 above)
Comment (keithwrightbos) [reply extensions/amp-a4a/0.1/amp-a4a.js]: Currently yes but any other FF network could elect to include RTC support

File: extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js (hunk 4 above)
Comment (keithwrightbos) [reply extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js]: This is done as part of building the ad url which is done for every refresh request.

File: extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js (hunk 5 above)
Comment (keithwrightbos) [reply extensions/amp-ad-network-adsense-impl/0.1/amp-ad-network-adsense-impl.js]: Yes, I later plan to include another ad request parameter indicating which state was used here

File: extensions/amp-a4a/0.1/amp-a4a.js (hunk 2 above)
Comment (keithwrightbos) [reply extensions/amp-a4a/0.1/amp-a4a.js]: Done

**Pull request merged**